internal/api/cluster.go

```diff
@@ -97,54 +198,92 @@ func handleGetCluster(c *Context, w http.ResponseWriter, r *http.Request) {
 	outputJSON(c, w, cluster)
 }
 
-// handleUpgradeCluster responds to PUT /api/clusters/{cluster}/kubernetes/{version}.
+// handleUpgradeCluster responds to PUT /api/clusters/{cluster}/kubernetes/{version}, upgrading
+// the cluster to the given Kubernetes version.
 func handleUpgradeCluster(c *Context, w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	clusterID := vars["cluster"]
 	version := vars["version"]
+	c.Logger = c.Logger.WithField("cluster", clusterID)
 
-	cluster, err := c.SQLStore.GetCluster(clusterID)
-	if err != nil {
-		c.Logger.WithField("error", err).Error("failed to query cluster")
-		w.WriteHeader(http.StatusInternalServerError)
+	cluster, status, unlockOnce := lockCluster(c, clusterID)
+	if status != 0 {
+		w.WriteHeader(status)
 		return
 	}
-	if cluster == nil {
-		w.WriteHeader(http.StatusNotFound)
+	defer unlockOnce()
+
+	switch cluster.State {
+	case model.ClusterStateStable:
+	case model.ClusterStateUpgradeRequested:
+	case model.ClusterStateUpgradeFailed:
+	default:
+		c.Logger.Warnf("unable to upgrade cluster while in state %s", cluster.State)
+		w.WriteHeader(http.StatusBadRequest)
 		return
 	}
 
-	err = c.Provisioner.UpgradeCluster(
-		clusterID,
-		version,
-	)
-	if err != nil {
-		c.Logger.WithField("error", err).Error("failed to upgrade cluster")
-		w.WriteHeader(http.StatusInternalServerError)
+	// TODO: Support something other than "latest".
+	if version != "latest" {
+		c.Logger.Warnf("unsupported kubernetes version %s", version)
+		w.WriteHeader(http.StatusBadRequest)
 		return
 	}
+
+	if cluster.State != model.ClusterStateUpgradeRequested {
+		cluster.State = model.ClusterStateUpgradeRequested
+
+		err := c.Store.UpdateCluster(cluster)
+		if err != nil {
+			c.Logger.WithError(err).Error("failed to mark cluster for upgrade")
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	unlockOnce()
+	c.Supervisor.Do()
+
+	w.WriteHeader(http.StatusAccepted)
 }
 
-// handleDeleteCluster responds to DELETE /api/clusters/{cluster}.
+// handleDeleteCluster responds to DELETE /api/clusters/{cluster}, beginning the process of
+// deleting the cluster.
 func handleDeleteCluster(c *Context, w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	clusterID := vars["cluster"]
+	c.Logger = c.Logger.WithField("cluster", clusterID)
 
-	cluster, err := c.SQLStore.GetCluster(clusterID)
-	if err != nil {
-		c.Logger.WithField("error", err).Error("failed to query cluster")
-		w.WriteHeader(http.StatusInternalServerError)
+	cluster, status, unlockOnce := lockCluster(c, clusterID)
+	if status != 0 {
+		w.WriteHeader(status)
 		return
 	}
-	if cluster == nil {
-		w.WriteHeader(http.StatusNotFound)
+	defer unlockOnce()
+
+	switch cluster.State {
```

Comment:
We should support additional states, such as deleting from a failed-creation state (or if not locked, from the creation requested state), but our kops+terraform provisioner chokes on that right now.